### ISSUE TITLE
The chunks don't match sometimes. Results may be qualitatively correct but quantitatively (slightly) off

### ISSUE THREAD

**Issue by 4gatepylon** [casparil/embedding-model-similarity#1]
# Different splitters use different tokenizers means that chunking isn't the same everywhere
I ran into this issue when trying to modify the repository for a class project and embeddings failed basic sanity tests (i.e. that the number of embeddings is the same between different models' embedded datasets because they should be 1:1 with the textual chunks). I think **chunks should be the same across embeddings** since otherwise the metrics aren't actually comparing the proper embeddings in some cases. Because the documents are small here it should be fine, but if the documents were very, very big, this would be a serious issue.

Notice how chunk ID (which should be unqiue by dataset) depends on the _index_ in that document after splitting with the splitter: https://github.com/casparil/embedding-model-similarity/blob/608f9ebdeddbe67f1b01147ff1f06c44b76110f4/owlergpt/utils/json_loader.py#L78. Notice that the splitter is not using the same tokenizer for all models. This means that if the document is long enough and you set the boundary just right, for some tokenizers there may be 1 chunk and for others there may be two (or two and three, etc...). This means that the chunks do not actually correspond to the same text. So IDs do not identify chunk text. This means that when you do the Jaccard similarity metric (etc...) the answer is sometimes a little bit incorrect (and artificially low). Also messes up the other similarity metrics because it means that you are not actually comparing the corresponding embeddings some of the time (i.e. for CKA).

Example of usage:
Getting indices (https://github.com/casparil/embedding-model-similarity/blob/608f9ebdeddbe67f1b01147ff1f06c44b76110f4/owlergpt/commands/evaluate_ds.py#L150) then immediately using those fetched embeddings for the analysis (                        queries1 = np.array(embeddings1["embeddings"])[query_inds1]) via `calculate_metric` into `_calculate_embed_metric` into i.e. `_cka` where the embeddings might not actually be for the same datapoints but are treated as if they are: https://github.com/casparil/embedding-model-similarity/blob/608f9ebdeddbe67f1b01147ff1f06c44b76110f4/owlergpt/utils/metrics.py#L164.

```python
if not match_dimension:
    queries1 = np.array(embeddings1["embeddings"])[query_inds1]
    queries2 = np.array(embeddings2["embeddings"])[query_inds2]

embeddings1 = torch.tensor(embeddings1["embeddings"])[idx1]
embeddings2 = torch.tensor(embeddings2["embeddings"])[idx2]
sim, fig = calculate_metric(embeddings1, embeddings2, queries1, queries2, metric, batch_size,
                            device, num_embeds, k, nn_metric, center, baseline)
```

Jupyter-copyable code to test this:
```python
import sys
from pathlib import Path
from tqdm import tqdm

# sys.path # does not include owlergpt
# Path.cwd() # commands
sys.path.append(Path.cwd().parent.parent.as_posix()) # Include folder within which owlergpt is located
from owlergpt.utils.json_loader import JSONDataset, collate_fn
from langchain.text_splitter import SentenceTransformersTokenTextSplitter, TokenTextSplitter

hf_models = [
    "Salesforce/SFR-Embedding-Mistral",
    "WhereIsAI/UAE-Large-V1",
    "BAAI/bge-base-en-v1.5",
    "BAAI/bge-large-en-v1.5",
    "BAAI/bge-small-en-v1.5",
    "intfloat/e5-base-v2",
    "intfloat/e5-large-v2",
    "intfloat/e5-small-v2",
    "thenlper/gte-base",
    "thenlper/gte-large",
    "thenlper/gte-small",
    "sentence-transformers/gtr-t5-base",
    "sentence-transformers/gtr-t5-large",
    "mixedbread-ai/mxbai-embed-large-v1",
    "sentence-transformers/sentence-t5-base",
    "sentence-transformers/sentence-t5-large",
]

print("Creating openai text splitters...")
openai_text_splitter_small = TokenTextSplitter(
    model_name="text-embedding-3-small",
    chunk_overlap=25,
    chunk_size=256
)
openai_text_splitter_large = TokenTextSplitter(
    model_name="text-embedding-3-large",
    chunk_overlap=25,
    chunk_size=256
)
print("Creating hf text splitters...")
text_splitters = [SentenceTransformersTokenTextSplitter(
        model_name=model_name,
        chunk_overlap=25,
        tokens_per_chunk=256,
    ) for model_name in tqdm(hf_models)
]
model_names = ["openai/text-embedding-3-small", "openai/text-embedding-3-large"] + hf_models


from tabulate import tabulate
from typing import List

# Initialize results dictionary
def visualize_table(model_names: List[str], sample_texts: List[str]):
    results = {model_name: [] for model_name in model_names}

    # Process each text with each splitter
    for i, text in enumerate(sample_texts, 1):
        # OpenAI splitters
        results["openai/text-embedding-3-small"].append(len(openai_text_splitter_small.split_text(text)))
        results["openai/text-embedding-3-large"].append(len(openai_text_splitter_large.split_text(text)))
        
        # HF splitters
        for model_name, splitter in zip(hf_models, text_splitters):
            results[model_name].append(len(splitter.split_text(text)))

    # Prepare table data
    headers = [f"Text {i}" for i in range(1, len(sample_texts) + 1)]
    table_data = [[model_name] + chunks for model_name, chunks in results.items()]

    # Print table
    print(tabulate(table_data, headers=["Model"] + headers, tablefmt="grid"))

hello_texts = ["hello " * i for i in range(1, 300, 10)]
visualize_table(model_names, hello_texts)
```

Notice that IDs are used i.e. for getting indices but IDs are no guarantee of mapping from the same text: https://github.com/casparil/embedding-model-similarity/blob/608f9ebdeddbe67f1b01147ff1f06c44b76110f4/owlergpt/utils/get_embedding_indices.py#L29

**Comment by casparil** [comment]
Thank you for pointing this out. I have updated the code in 0fe55ebe9b6f3da2035c3a5eb0b527ca312be0cf and added the option to specify a fixed tokenizer using the `VECTOR_SEARCH_TEXT_SPLITTER` variable (see `.env.example` for reference).

While fixing the tokenizer ensures that models embed the same text, it might lead to issues when setting the chunk size close to a model's maximum tokens. If the embedding model's tokenizer is different from the fixed tokenizer, the model tokenizer might need more tokens to encode the same text causing it to exceed the model's maximum token size. However, in our specific setting where we chose a chunk size of 256 and models supporting at least 512 tokens, fixing the tokenizer worked fine.

To illustrate how changing from model specific tokenizers to a fixed tokenizer (using `e5-small`) affects similarity, see the image below:

![Image](https://github.com/user-attachments/assets/fc904915-758a-4533-917c-28f24e19a910)

As the `e5`, `bge` and `gte` models as well as `UAE` and `mxbai` all use a `BertTokenizer`, similarities between these models remain unaffected. For the remaining models, similarities increase on average (+0.017).
